src/evaluator.jl

```diff
@@ -99,34 +116,36 @@ end
 grad_config!(g, m::LinearACEModel, V::ProductEvaluator, cfg::AbstractConfiguration) = 
       grad_config!(g, V, cfg)
 
+
 # compute one site energy
 function grad_config!(g, V::ProductEvaluator, cfg::AbstractConfiguration)
    basis1p = V.pibasis.basis1p
    _real = V.pibasis.real
    A = acquire_B!(V.pibasis.basis1p, cfg)
    dA = acquire_dB!(V.pibasis.basis1p, cfg)
    dAAdA = _acquire_dAAdA!(V.pibasis)
-
+   
    # stage 1: precompute all the A values
    evaluate_ed!(A, dA, basis1p, cfg)
 
    # stage 2: compute the coefficients for the ∇A_{klm} = ∇ϕ_{klm}
-   dAco = zeros(eltype(dAAdA), length(A)) # tmpd.dAco  # TODO: ALLOCATION 
    c̃ = V.coeffs
+   dAco =  _alloc_dAco(dAAdA, A, c̃) # tmpd.dAco  # TODO: ALLOCATION 
    spec = V.pibasis.spec
+
    fill!(dAco, zero(eltype(dAco)))
    @inbounds for iAA = 1:length(spec)
       _AA_local_adjoints!(dAAdA, A, spec.iAA2iA, iAA, spec.orders[iAA], _real)
       @fastmath for t = 1:spec.orders[iAA]
-         dAco[spec.iAA2iA[iAA, t]] += dAAdA[t] * complex(c̃[iAA])
+         dAco[spec.iAA2iA[iAA, t]] += dAAdA[t] * complex(c̃[iAA]) #trying to avoid using .* and complex.()
       end
    end
 
    # stage 3: get the gradients
    fill!(g, zero(eltype(g)))
-   for iX = 1:length(cfg)
-      @inbounds @fastmath for iA = 1:length(basis1p)
-         g[iX] += _real(dAco[iA] * dA[iA, iX])
+   for iP = 1:length(c̃[1]), iX = 1:length(cfg)
+      for iA = 1:length(basis1p)
+         g[iX, iP] += _real(dAco[iA][iP] * dA[iA, iX])
```

Comment:
I assume what matters is how we create our rrules since we can't use raw ChainRules or Zygote anyways. However, I think we should do whatever is more convenient for our other use cases. Will we ever use grad_config without Zygote?

Reply:
I expect yes, very often.

Reply:
Then it might be worth discussing this a little more. I don't think it will directly affect Zygote, but still worth discussing.

Reply:
next PR I think.